nsmdp/nsmdp.go

```diff
@@ -20,89 +20,123 @@
-		stop:         make(chan interface{}),
-		updatedevs:   make(chan *NSMDevice, 10),
-	}
-	for i := uint(0); i < initDeviceCount; i++ {
-		dev := &NSMDevice{
-			Device: &pluginapi.Device{
-				ID:     "NSM_" + string(i),
-				Health: pluginapi.Healthy,
-			},
-			token: generateToken(),
-		}
-		n.updatedevs <- dev
+func startClientServer(client *nsmSocket, logger logging.PluginLogger) {
+	logger.Infof(" startClientServer was called with socket: %s", client.socketPath)
+	listenEndpoint := client.socketPath
+	fi, err := os.Stat(listenEndpoint)
+	if err == nil && (fi.Mode()&os.ModeSocket) != 0 {
+		os.Remove(listenEndpoint)
+	}
+	if err != nil && !os.IsNotExist(err) {
+		logger.Infof(" failure stat of socket file %s with error: %+v", client.socketPath, err)
+		client.allocated = false
+		return
 	}
-	return n
-}
 
-func generateToken() string {
-	b := make([]byte, 16)
-	rand.Read(b)
-	return fmt.Sprintf("%x", b)
-}
+	unix.Umask(0077)
+	sock, err := net.Listen("unix", listenEndpoint)
+	if err != nil {
+		logger.Infof(" failure to listen on socket %s with error: %+v", client.socketPath, err)
+		client.allocated = false
+		return
+	}
+	grpcServer := grpc.NewServer([]grpc.ServerOption{}...)
+
+	logger.Infof(" Starting Client gRPC server listening on socket: %s", serverSock)
+	go grpcServer.Serve(sock)
 
-func (n *NSMDevicePlugin) Stop() error {
-	err := n.DevicePlugin.Stop()
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	conn, err := dial(ctx, listenEndpoint)
 	if err != nil {
-		return err
+		logger.Infof("failure to communicate with the socket %s with error: %+v", client.socketPath, err)
+		client.allocated = false
 	}
-	close(n.stop)
-	return nil
+	defer conn.Close()
+	defer cancel()
+	logger.Infof(" Client Server socket: %s is operational", listenEndpoint)
 
+	// Wait for shutdown
+	select {
+	case <-client.stopChannel:
+		logger.Infof(" Server for socket %s received shutdown request", client.socketPath)
+	}
+	client.allocated = false
+	client.stopChannel <- true
 }
 
-func (n *NSMDevicePlugin) Allocate(ctx context.Context, reqs *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+func (n *nsmClientEndpoints) Allocate(ctx context.Context, reqs *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	n.logger.Info(" Allocate was called.")
 	responses := pluginapi.AllocateResponse{}
 	for _, req := range reqs.ContainerRequests {
 		var mounts []*pluginapi.Mount
 		for _, id := range req.DevicesIDs {
-			_, ok := n.devs[id]
-			if !ok {
-				return nil, fmt.Errorf("invalid allocation request: unknown device: %s", id)
-			}
-			mount := &pluginapi.Mount{
-				ContainerPath: SocketBaseDir + id,
-				HostPath:      SocketBaseDir + id,
+			if _, ok := n.nsmSockets[id]; ok {
+				if n.nsmSockets[id].allocated {
+					// Socket has been previsously used, since we did noy get notification from
+					// kubelet when POD using this socket went down, gRPC client's server
+					// needs to be stopped.
```

Comment:
This is a good catch, thanks for catching and handling this case where the soccer is re-used!